# HG changeset patch
# User Rik <rik@octave.org>
# Date 1443475629 25200
#      Mon Sep 28 14:27:09 2015 -0700
# Node ID 3aea4200da40e47f94802a5307d71a5515b29ec8
# Parent  88459dd6780fb04562a7f7f203ecc14da8e7b5f8
# Parent  bbe6a846d8ac938b17764dc5175407c2e7714268
maint: Periodic merger of stable to default.

diff --git a/libgui/graphics/Canvas.cc b/libgui/graphics/Canvas.cc
--- a/libgui/graphics/Canvas.cc
+++ b/libgui/graphics/Canvas.cc
@@ -626,19 +626,30 @@ Canvas::canvasMousePressEvent (QMouseEve
                                 Utils::figureSelectionType (event, isdblclick), false);
 
           updateCurrentPoint (figObj, obj, event);
 
           gh_manager::post_callback (figObj.get_handle (),
                                      "windowbuttondownfcn",
                                      button_number (event));
 
-          gh_manager::post_callback (currentObj.get_handle (),
-                                     "buttondownfcn",
-                                     button_number (event));
+          if (currentObj.get ("buttondownfcn").is_empty ())
+            {
+              graphics_object parentObj = 
+                gh_manager::get_object (currentObj.get_parent ());
+
+              if (parentObj.valid_object () && parentObj.isa ("hggroup"))
+                gh_manager::post_callback (parentObj.get_handle (),
+                                           "buttondownfcn",
+                                           button_number (event));
+            }
+          else
+            gh_manager::post_callback (currentObj.get_handle (),
+                                       "buttondownfcn",
+                                       button_number (event));
 
           if (event->button () == Qt::RightButton)
             ContextMenu::executeAt (currentObj.get_properties (),
                                     event->globalPos ());
           break;
 
         case TextMode:
           {
diff --git a/libgui/graphics/gl-select.cc b/libgui/graphics/gl-select.cc
--- a/libgui/graphics/gl-select.cc
+++ b/libgui/graphics/gl-select.cc
@@ -202,21 +202,21 @@ opengl_selector::render_text (const std:
 #endif
 }
 
 void
 opengl_selector::draw_image (const image::properties& props)
 {
   Matrix xd = props.get_xdata ().matrix_value ();
   octave_idx_type nc = props.get_cdata ().columns ();
-  double x_pix_size = (xd(1) - xd(0)) / (nc - 1);
+  double x_pix_size = (nc == 1 ? 1 : (xd(1) - xd(0)) / (nc - 1));
 
   Matrix yd = props.get_ydata ().matrix_value ();
   octave_idx_type nr = props.get_cdata ().rows ();
-  double y_pix_size = (yd(1) - yd(0)) / (nr - 1);
+  double y_pix_size = (nr == 1 ? 1 : (yd(1) - yd(0)) / (nr - 1));
 
   ColumnVector p1(3, 0.0), p2(3, 0.0), p3(3, 0.0), p4(3, 0.0);
   p1(0) = xd(0) - x_pix_size/2;
   p1(1) = yd(0) - y_pix_size/2;
 
   p2(0) = xd(1) + x_pix_size/2;
   p2(1) = yd(0) - y_pix_size/2;
 
diff --git a/libinterp/corefcn/mex.h b/libinterp/corefcn/mex.h
--- a/libinterp/corefcn/mex.h
+++ b/libinterp/corefcn/mex.h
@@ -59,16 +59,18 @@ typedef int bool;
 /* -V4 stuff */
 #if defined (V4)
 #define Matrix mxArray
 #define REAL mxREAL
 #endif
 
 #define mxMAXNAME 64
 
+#include <stdint.h>
+
 #include "mexproto.h"
 
 #if defined (__cplusplus)
 extern "C" {
 #endif
 
 #if defined (V4)
 void mexFunction (int nlhs, mxArray* plhs[], int nrhs, mxArray *prhs[]);
diff --git a/liboctave/system/file-ops.cc b/liboctave/system/file-ops.cc
--- a/liboctave/system/file-ops.cc
+++ b/liboctave/system/file-ops.cc
@@ -49,16 +49,20 @@ extern "C" {
 #include "oct-env.h"
 #include "oct-locbuf.h"
 #include "oct-passwd.h"
 #include "pathlen.h"
 #include "quit.h"
 #include "singleton-cleanup.h"
 #include "str-vec.h"
 
+#if (defined (OCTAVE_HAVE_WINDOWS_FILESYSTEM) && ! defined (OCTAVE_HAVE_POSIX_FILESYSTEM))
+#include <algorithm>
+#endif
+
 file_ops *file_ops::instance = 0;
 
 bool
 file_ops::instance_ok (void)
 {
   bool retval = true;
 
   if (! instance)
@@ -733,14 +737,19 @@ octave_canonicalize_file_name (const std
   char *tmp = gnulib::canonicalize_file_name (name.c_str ());
 
   if (tmp)
     {
       retval = tmp;
       free (tmp);
     }
 
+#if (defined (OCTAVE_HAVE_WINDOWS_FILESYSTEM) && ! defined (OCTAVE_HAVE_POSIX_FILESYSTEM))
+  // Canonical Windows file separator is backslash.
+  std::replace (retval.begin (), retval.end (), '/', '\\');
+#endif
+
   if (retval.empty ())
     msg = gnulib::strerror (errno);
 
   return retval;
 }
 
diff --git a/scripts/plot/appearance/annotation.m b/scripts/plot/appearance/annotation.m
--- a/scripts/plot/appearance/annotation.m
+++ b/scripts/plot/appearance/annotation.m
@@ -95,28 +95,30 @@
 ## @qcode{"head1length"}/@qcode{"head2length"},
 ## @qcode{"head1width"}/@qcode{"head2width"}, etc.  The index 1 marks the
 ## properties of the arrowhead at the first point in @var{x} and @var{y}
 ## coordinates.
 ##
 ## @item @qcode{"textarrow"}
 ## Construct an arrow with a text label at the opposite end from the arrowhead.
 ##
+## Use the "string" property to change the text string.
 ## The line and the arrowhead can be customized as for arrow annotations, and
 ## the text can be customized using the same properties as @code{text} graphics
 ## objects.  Note, however, that some text property names are prefixed with
 ## "text" to distinguish them from arrow properties:
 ## @qcode{"textbackgroundcolor"}, @qcode{"textcolor"},
 ## @qcode{"textedgecolor"}, @qcode{"textlinewidth"},
 ## @qcode{"textmargin"}, @qcode{"textrotation"}.
 ##
 ## @item @qcode{"textbox"}
 ## Construct a box with text inside.  @var{pos} specifies the
 ## @qcode{"position"} property of the annotation.
 ##
+## Use the "string" property to change the text string.
 ## You may use @qcode{"backgroundcolor"}, @qcode{"edgecolor"},
 ## @qcode{"linestyle"}, and @qcode{"linewidth"} properties to customize
 ## the box background color and edge appearance.  A limited set of @code{text}
 ## objects properties are also available; Besides @qcode{"font@dots{}"}
 ## properties, you may also use @qcode{"horizontalalignment"} and
 ## @qcode{"verticalalignment"} to position the text inside the box.
 ##
 ## Finally, the @qcode{"fitboxtotext"} property controls the actual extent of
diff --git a/src/mkoctfile.in.cc b/src/mkoctfile.in.cc
--- a/src/mkoctfile.in.cc
+++ b/src/mkoctfile.in.cc
@@ -780,16 +780,17 @@ main (int argc, char **argv)
                 << "mkoctfile: no way to link stand-alone executable file"
                 << std::endl;
               return 1;
             }
         }
       else
         {
           std::string cmd = vars["DL_LD"] + " "
+                            + vars["ALL_CXXFLAGS"] + " "
                             + vars["DL_LDFLAGS"] + " "
                             + pass_on_options
                             + " -o " + octfile + " "
                             + objfiles + " " + libfiles + " "
                             + ldflags + " "
                             + vars["LFLAGS"] + " -loctinterp -loctave "
                             + vars["OCT_LINK_OPTS"] + " "
                             + vars["OCT_LINK_DEPS"];
